### PR TITLE
Fix discrepancies between `FIRE` optimisations in `ASE` and `torch-sim` 

### PR DESCRIPTION
**Pull request by t-reents**
It seems that the remaining deviations could be fixed. This branch is already based on the changes in #201, I'll do the rebase once it is merged.

* I reduced the number of "inner" updates in the testing script to `1`, in order to get the precise number of convergence steps
* In the pos-only case, two small fixes were necessary that led to a wrong (or better to say different) `dt`
* The Frechet Cell optimization also required some additional changes, i.e., correctly scaling the positions with respect to the deformation gradient. See https://gitlab.com/ase/ase/-/blob/master/ase/filters.py?ref_type=heads#L440-441 and https://gitlab.com/ase/ase/-/blob/master/ase/filters.py?ref_type=heads#L405-406
* The previously (#199) introduced rescaling of the positions with respect to the new cell was not necessary (it seems that those were anyway overwritten in ASE: https://gitlab.com/ase/ase/-/blob/master/ase/filters.py?ref_type=heads#L434-441), so I removed it again
* Finally, I adapted the `generate_force_convergence_fn` function to optionally also include the `cell_forces` in the convergence check, which is done in `ASE` and therefore caused the differing numbers of convergence steps

fig1

fig2

fig3

### CODE REVIEW

File: torch_sim/optimizers.py [hunk 1]
```diff
@@ -1554,10 +1556,17 @@ def _ase_fire_step(  # noqa: C901, PLR0915
         old_row_vector_cell = state.row_vector_cell.clone()
 
     dr_scaling_atom = torch.sqrt(dr_scaling_batch)[state.batch].unsqueeze(-1)
+
     dr_atom = torch.where(
         dr_scaling_atom > max_step, max_step * dr_atom / (dr_scaling_atom + eps), dr_atom
     )
-    state.positions = state.positions + dr_atom
+
+    if is_cell_optimization:
```
Comment (CompRhys) [inline torch_sim/optimizers.py:1558]: this seems strange to me? does this logic not add the cartesian dr to the fractional space positions?

File: torch_sim/optimizers.py [hunk 2]
```diff
@@ -1590,11 +1599,10 @@ def _ase_fire_step(  # noqa: C901, PLR0915
         transform_matrix_batch = torch.bmm(
             inv_old_cell_batch, current_new_row_vector_cell
         )
-        atom_specific_transform = transform_matrix_batch[state.batch]
-        scaled_positions = torch.bmm(
-            state.positions.unsqueeze(1), atom_specific_transform
-        ).squeeze(1)
-        state.positions = scaled_positions
+
```
Comment (CompRhys) [inline torch_sim/optimizers.py:1589]: should lines 1596-1602 not be deleted?

File: torch_sim/optimizers.py (hunk 1 above)
Comment (CompRhys) [reply torch_sim/optimizers.py:1558]:
```python
        positions = (
            torch.linalg.solve(
                cur_deform_grad[state.batch], (state.positions+dr_atom).unsqueeze(-1)
            ).squeeze(-1)
        )
        state.positions = torch.bmm(
            positions.unsqueeze(1), F_new[state.batch].transpose(-2, -1)
        ).squeeze(1)
```

why would this not be the right thing to do in terms of trying to match your operations here?

**Comment by t-reents** [comment]
I would have to look into the failing test in more detail. As a first step, I took the `rattled_sio2` and manually inspected the whole relaxation trajectory (I did the same for the other debugging as well, i.e., a bunch of print statements comparing all the tensors etc. during relaxation). The final optimized structure as well as the intermediate states seem to agree very well. 

Still, the test is failing, so I'd have to understand what exactly is happening there. Nonetheless, based on my previous comment, those deviations might be related to the logic of the test and not necessarily to actual differences in the optimizations. I'll look into it and give an update here.


**Update:** I have to stop for now, but there is indeed a problem with the logic of the tests/the loop logic in `torch_sim.runners.optimize`. I'll push a fix tomorrow or Monday at the lastest. The failing of the tests for some of the examples is related to a different number of iterations that are performed, while for others it's the reinitialization (the init function is always called again, e.g. resetting the velocities to 0, while ASE keeps the previous state). When adjusting those aspects, I the test does pass locally.

File: torch_sim/optimizers.py (hunk 1 above)
Comment (t-reents) [reply torch_sim/optimizers.py:1558]: > this seems strange to me? does this logic not add the cartesian dr to the fractional space positions?

No, it doesn't scale between fractional/cartesian coordinates. Based on the comments in ASE, this step is done to consider the positions in the original cell. The `dr_atom` is applied to those coordinates and the new cell deformation is applied afterwards. This is also briefly mentioned in the documentation of the `UnitCellFilter` (https://wiki.fysik.dtu.dk/ase/ase/filters.html#ase.filters.UnitCellFilter):

```
Degrees of freedom are the positions in the original undeformed cell, plus the 
deformation tensor (extra 3 “atoms”). This gives forces consistent with numerical 
derivatives of the potential energy with respect to the cell degreees of freedom.
```

> why would this not be the right thing to do in terms of trying to match your operations here?

Again, to my understanding, you want to consider the positions in the undeformed cell and apply the `dr_atom` to those. That's why I apply the `linalg.solve` first and then add the `dr_atom`, and not the other way around.

**Comment by t-reents** [comment]
The test is now passing, at least locally. As mentioned earlier, one of the issues was the reinitializations between the different steps that are tested. To fix it, I added an if statement for now that checks if the state is already of the correct type.

Moreover, as described in the commit message, the current logic of the autobatching caused a different number of iterations, since the structures are set as converged when the maximum number of swap attempts is reached and not necessarily the maximum number of iterations (which led to different number of steps compared to ASE).

**Comment by CompRhys** [comment]
@t-reents thank you for your careful work here!
